NotificationHubs/src/com/windowsazure/messaging/Registration.java

```diff
@@ -15,330 +20,381 @@
-		digester.addSetNext("*/AdmRegistrationDescription", "addRegistration");
-		digester.addSetNext("*/AdmTemplateRegistrationDescription", "addRegistration");
-		digester.addSetNext("*/BaiduRegistrationDescription", "addRegistration");
-		digester.addSetNext("*/BaiduTemplateRegistrationDescription", "addRegistration");
-	}
-
-	public static class RegistrationCreationFactory implements
-			ObjectCreationFactory<Object> {
-		private Digester digester;
-
-		@Override
-		public Object createObject(Attributes attributes) throws Exception {
-			if ("WindowsRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new WindowsRegistration();
-			}
-			if ("WindowsTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new WindowsTemplateRegistration();
-			}
-			if ("AppleRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new AppleRegistration();
-			}
-			if ("AppleTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new AppleTemplateRegistration();
-			}
-			if ("GcmRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new GcmRegistration();
-			}
-			if ("GcmTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new GcmTemplateRegistration();
-			}
-			if ("FcmRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new FcmRegistration();
-			}
-			if ("FcmTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new FcmTemplateRegistration();
-			}
-			if ("MpnsRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new MpnsRegistration();
-			}
-			if ("MpnsTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new MpnsTemplateRegistration();
-			}
-			if ("AdmRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new AdmRegistration();
-			}
-			if ("AdmTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new AdmTemplateRegistration();
-			}
-			if ("BaiduRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new BaiduRegistration();
-			}
-			if ("BaiduTemplateRegistrationDescription".equals(attributes
-					.getValue("i:type"))) {
-				return new BaiduTemplateRegistration();
-			}
-			throw new UnsupportedOperationException("unknown type: "
-					+ attributes.getValue("i:type"));
-		}
-
-		@Override
-		public Digester getDigester() {
-			return digester;
-		}
-
-		@Override
-		public void setDigester(Digester digester) {
-			this.digester = digester;
-		}
-	}
+    protected String registrationId;
+    protected Set<String> tags = new HashSet<>();
+    protected String etag;
+    protected Date expirationTime;
+
+    private static final AtomicReference<Digester> singleRegParser = new AtomicReference<>();
```

Comment:
Same question regarding the difference between ThreadLocal and AtomicReference.

Reply:
Changed all to use Atomic Reference instead of ThreadLocal